### PR TITLE
fix RuleStatus update fail

### LINKED ISSUE #2920: The subresource Status of the routing module crd resource Rule failed to update
**What happened**:
When I am testing custom sending messages，I found that the number of times the message was forwarded was always 0，The resource field is Status.SuccessMessages
**What you expected to happen**:
When I send a message, the Rule.Status.SuccessMessages increases
**How to reproduce it (as minimally and precisely as possible)**:
Follow this [tutorial，](https://kubeedge.io/en/docs/developer/custom_message_deliver/#how-to-deliver-custom-messages)
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.19.1
- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`):1.7.1

Read the source code to find the cause of this [upstream.go-updateRuleStatus](https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/edgecontroller/controller/upstream.go)
```go
rule, err = uc.crdClient.RulesV1().Rules(namespace).Get(context.Background(), ruleID, metaV1.GetOptions{})
			if err != nil {
				klog.Warningf("message: %s process failure, get rule with error: %s, namespaces: %s name: %s", msg.GetID(), err, namespace, ruleID)
				continue
			}
			content, ok := msg.Content.(routerrule.ExecResult)
			if !ok {
				klog.Warningf("message: %s process failure, get rule content with error: %s, namespaces: %s name: %s", msg.GetID(), err, namespace, ruleID)
				continue
			}
			if content.Status == "SUCCESS" {
				rule.Status.SuccessMessages++
			}
			if content.Status == "FAIL" {
				rule.Status.FailMessages++
				errSlice := make([]string, 0)
				rule.Status.Errors = append(errSlice, content.Error.Detail)
			}
			newStatus := &rulesv1.RuleStatus{}
			body, err := json.Marshal(newStatus)
			if err != nil {
				klog.Warningf("message: %s process failure, content marshal err: %s", msg.GetID(), err)
				continue
			}
			var data []byte = []byte(body)
			_, err = uc.crdClient.RulesV1().Rules(namespace).Patch(context.Background(), ruleID, controller.MergePatchType, data, metaV1.PatchOptions{})
			if err != nil {
				klog.Warningf("message: %s process failure, update ruleStatus failed with error: %s, namespace: %s, name: %s", msg.GetID(), err, namespace, ruleID)
			} else {
				klog.Infof("UpdateRulestatus successfully!")
			}

```
There are two problems here. A newStatus is only initialized. Second, Rule.Status is a sub-resource and needs to be enabled in crd yaml

### PR DESCRIPTION
**Pull request by xujiaja**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Created a new structure of RuleStatus and then assign the value of Rule'status.

**Which issue(s) this PR fixes**:

Fixes #2920 

**Special notes for your reviewer**:
none

**Does this PR introduce a user-facing change?**:
none

```release-note

```

### CODE REVIEW

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 1]
```diff
@@ -220,6 +220,11 @@ func (uc *UpstreamController) dispatchMessage() {
 		}
 	}
 }
+
+type RuleStatus struct {
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/controller/upstream.go:224]: Why do we need this struct?

**Review by fisherxu — COMMENTED** [review]
please fix the failed DCO ci test(signoff the commit).

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 1 above)
Comment (xujiaja) [reply cloud/pkg/edgecontroller/controller/upstream.go:224]: k8s patch RuleStatus must specify the object of status, may be t this structure is placed in the wrong place

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 2]
```diff
@@ -257,7 +262,8 @@ func (uc *UpstreamController) updateRuleStatus() {
 				errSlice := make([]string, 0)
 				rule.Status.Errors = append(errSlice, content.Error.Detail)
 			}
-			newStatus := &rulesv1.RuleStatus{}
+			newStatus := &RuleStatus{}
+			newStatus.Status = rule.Status
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/controller/upstream.go:266]: If we always patch the empty status, the count will always 0. We should patch the rule's real status.

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 3]
```diff
@@ -257,7 +262,8 @@ func (uc *UpstreamController) updateRuleStatus() {
...
 			body, err := json.Marshal(newStatus)
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/controller/upstream.go:267]: Thanks for the fixing, and has any test here? Is the status worked as desired? For the patch func, the data should be patch bytes created through olddata and newdata. 

Ref: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go#L346

**Comment by xujiaja** [comment]
The k8s cluster failed to initialize due to timeout, please re-execute CI

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 2 above)
Comment (fisherxu) [inline cloud/pkg/edgecontroller/controller/upstream.go:266]: Could we use update status as below?
```go
if content.Status == "SUCCESS" {
				rule.Status.SuccessMessages++
			}
			if content.Status == "FAIL" {
				rule.Status.FailMessages++
				errSlice := make([]string, 0)
				rule.Status.Errors = append(errSlice, content.Error.Detail)
			}

			_, err = uc.crdClient.RulesV1().Rules(namespace).UpdateStatus(context.Background(), rule, metaV1.UpdateOptions{})
			if err != nil {
				klog.Warningf("message: %s process failure, update ruleStatus failed with error: %s, namespace: %s, name: %s", msg.GetID(), err, namespace, ruleID)
			} else {
				klog.Infof("UpdateRulestatus successfully!")
			}
```

**Comment by fisherxu** [comment]
And please rebase to master branch to fix the failed ci tests :)

**Pull request closed without merging**